src/reducers/sources.js

```diff
@@ -69,14 +72,14 @@ function update(state = State(), action: Action) : Record<SourcesState> {
       });
 
     case "CLOSE_TAB":
-      availableTabs = removeSourceFromTabList(state.tabs, action.id);
+      availableTabs = removeSourceFromTabList(state, state.tabs, action.id);
```

Comment:
@jasonLaster 

I think for restoring a tab the variables that need to be updated actually reside in `case "SELECT_SOURCE":`, so we don't need to add anything to this case. However, just executing `case "SELECT_SOURCE":` is not  enough because the source text would not be loaded, so we have to use `dispatch(selectSource(source.id));` in `src/actions/sources.js`. `selectSource(source.id)` would execute `case "SELECT_SOURCE` and load the source text.